File: 5KYU/Tic-Tac-Toe Checker/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class TicTacToe
{
    private static bool IsFinished(int[,] board) 
        => board.Cast<int>().All(xy => xy != 0);

    private static readonly List<List<(int X, int Y)>> winningStates
        = new List<List<(int X, int Y)>>
            {  new List<(int X, int Y)>{(0, 0), (0, 1), (0, 2) },
               new List<(int X, int Y)>{(1, 0), (1, 1), (1, 2) },
               new List<(int X, int Y)>{(2, 0), (2, 1), (2, 2) },
               new List<(int X, int Y)>{(0, 0), (1, 0), (2, 0) },
               new List<(int X, int Y)>{(0, 1), (1, 1), (2, 1) },
               new List<(int X, int Y)>{(0, 2), (1, 2), (2, 2) },
               new List<(int X, int Y)>{(0, 0), (1, 1), (2, 2) },
               new List<(int X, int Y)>{(0, 2), (1, 1), (2, 0) }
            };
    
    public int IsSolved(int[,] board)
    {
        // check if there is a winner.
        foreach (var i in Enumerable.Range(1, 2))
        {
            foreach (var ws in winningStates)
                if (ws.Select(xy => board[xy.X, xy.Y]).All(n => n == i))
                    return i;
        }
        // no winners, so check if the board is finished, if it its a draw.
        if (board.Cast<int>().All(xy => xy != 0))
             return 0;
        
        // else the game is not finished.
        return -1;
    }
}

File: 5KYU/Tic-Tac-Toe Checker/solution.go
package kata

func Winner(board [3][3]int, state [][]int) int {
	s1x := state[0][0]
	s1y := state[0][1]
	s2x := state[1][0]
	s2y := state[1][1]
	s3x := state[2][0]
	s3y := state[2][1]

	if board[s1x][s1y] == board[s2x][s2y] && board[s1x][s1y] == board[s3x][s3y] &&
		board[s1x][s1y] != 0 {
		return board[s1x][s1y]
	}
	return -1
}

func BoardComplete(board [3][3]int) bool {
	for x := 0; x < 3; x++ {
		for y := 0; y < 3; y++ {
			if board[x][y] == 0 {
				return false
			}
		}
	}
	return true
}

func IsSolved(board [3][3]int) int {

	winningStates := [][][]int{
		{{0, 0}, {0, 1}, {0, 2}},
		{{1, 0}, {1, 1}, {1, 2}},
		{{2, 0}, {2, 1}, {2, 2}},
		{{0, 0}, {1, 0}, {2, 0}},
		{{0, 1}, {1, 1}, {2, 1}},
		{{0, 2}, {1, 2}, {2, 2}},
		{{0, 0}, {1, 1}, {2, 2}},
		{{0, 2}, {1, 1}, {2, 0}}}

	for _, winningState := range winningStates {
		winner := Winner(board, winningState)
		if winner != -1 {
			return winner
		}
	}

	if BoardComplete(board) {
		return 0
	} else {
		return -1
	}
}
